lib/segment/src/id_tracker/mutable_id_tracker.rs

```diff
@@ -368,20 +328,94 @@ fn versions_path(segment_path: &Path) -> PathBuf {
     segment_path.join(FILE_VERSIONS)
 }
 
-fn load_mappings(
-    mappings_path: &Path,
-    deleted: &mut BitVec,
-    internal_to_external: &mut Vec<PointIdType>,
-    external_to_internal_num: &mut BTreeMap<u64, PointOffsetType>,
-    external_to_internal_uuid: &mut BTreeMap<Uuid, PointOffsetType>,
+/// Store new mapping changes, appending them to the given file
+fn store_mapping_changes(mappings_path: &Path, changes: Vec<MappingChange>) -> OperationResult<()> {
+    // Open file in append mode to write new changes to the end
+    let file = File::options()
+        .create(true)
+        .append(true)
+        .open(mappings_path)?;
+    let mut writer = BufWriter::new(file);
+
+    write_mapping_changes(&mut writer, changes).map_err(|err| {
+        OperationError::service_error(format!(
+            "Failed to persist ID tracker point mappings ({}): {err}",
+            mappings_path.display(),
+        ))
+    })?;
+
+    // Explicitly fsync file contents to ensure durability
+    let file = writer.into_inner().unwrap();
+    file.sync_all().map_err(|err| {
+        OperationError::service_error(format!("Failed to fsync ID tracker point mappings: {err}"))
+    })?;
+
+    Ok(())
+}
+
+/// Serializes pending point mapping changes into the given writer
+///
+/// ## File format
+///
+/// All entries have a variable size and are simply concatenated. Each entry has a 1-byte header
+/// which specifies the change type and implies the length of the entry.
+///
+/// See [`read_entry`] and [`write_entry`] for more details.
+fn write_mapping_changes<W: Write>(
+    mut writer: W,
+    changes: Vec<MappingChange>,
 ) -> OperationResult<()> {
+    for change in changes {
+        write_entry(&mut writer, change)?;
+    }
+
+    // Explicitly flush writer to catch IO errors
+    writer.flush()?;
+
+    Ok(())
+}
+
+/// Load point mappings from the given file
+fn load_mappings(mappings_path: &Path) -> OperationResult<PointMappings> {
     let mappings_file = File::open(mappings_path)?;
-    let mappings_reader = std::io::BufReader::new(mappings_file);
+    let mappings_reader = BufReader::new(mappings_file);
+    read_mappings(mappings_reader)
+}
+
+/// Iterate over mapping changes from the given reader
+///
+/// The iterator ends when the end of the file is reached, or when an error occurred.
+///
+/// ## Error
+///
+/// An error item is returned if reading a mapping change fails due to malformed data. Then the
+/// iterator will not produce any more items.
+fn read_mappings_iter<R: Read>(
+    mut reader: R,
+) -> impl FusedIterator<Item = OperationResult<MappingChange>> {
+    // Keep reading until end of file or error
+    std::iter::from_fn(move || match read_entry(&mut reader) {
+        Ok(entry) => Some(Ok(entry)),
+        // Done reading if end of file is reached
+        Err(err) if err.kind() == io::ErrorKind::UnexpectedEof => None,
```

Comment:
What would happen if we have a corrupted file, in which we ignored last entry error, in case we continue to write new records into it?

Reply:
Thanks! That would indeed be a problem. I had it on my task list, but forgot it in this PR.

I'll try to add some logic to truncate it.

Reply:
I'll handle this in a separate PR, as it isn't trivial to do and I don't want to make this PR more complicated. I've added it in the [tracking issue](https://github.com/qdrant/qdrant/issues/6157).